python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,501 @@
+            isGood = np.copy(isGood_init)
+            isGood[isGood_init == 1] = isGood_init[isGood_init == 1]*~large_outliers_selection
+
+            # Replace all unsuccessful measurments wih nan
+            x_measurments[~isGood] = np.nan
+            y_measurments[~isGood] = np.nan
+
+            # Specify the number of breaks in the data
+            if np.sum(np.diff(isGood)) == 1:
+                number_of_breaks = 1
+            else:
+                number_of_breaks = 2
+
+            if type_of_run == 'theta':
+                poly_order_inter = order_theta_inter
+                poly_order_extra = order_theta_extra
+            if type_of_run == 'phi':
+                poly_order_inter = order_phi_inter
+                poly_order_extra = order_phi_extra
+            # Create prediction for where the points should be, when you do not see points
+            # If you do not see points from both sides of the black dots, do simpler extrapolation
+            # If you see points from both sides of the black dots, do more complex interpolation
+            if number_of_breaks == 1:
+                deg2_fit_to_single_point_y, residuals_y =\
+                    self.poly_fit(n_obs, y_measurments, isGood, poly_order_extra)
+                deg2_fit_to_single_point_x, residuals_x =\
+                    self.poly_fit(n_obs, x_measurments, isGood, poly_order_extra)
+            else:
+                deg2_fit_to_single_point_y, residuals_y =\
+                    self.poly_fit(n_obs, y_measurments, isGood, poly_order_inter)
+                deg2_fit_to_single_point_x, residuals_x =\
+                    self.poly_fit(n_obs, x_measurments, isGood, poly_order_inter)
+
+            poly_deg_x = np.poly1d(deg2_fit_to_single_point_x)
+            poly_deg_y = np.poly1d(deg2_fit_to_single_point_y)
+
+            # Cleaning, to exclude the results where the interpolation results make no sense
+            # If cobra has more than 5 observations hidden, calculate how much distance
+            # we predicted that the cobra has moved during the `predicted` movement,
+            # i.e., while it was not observed. This will be used
+            # below to remove prediction that make no sense
+            x_predicted = poly_deg_x(np.arange(0, n_obs))[~isGood]
+            y_predicted = poly_deg_y(np.arange(0, n_obs))[~isGood]
+            if np.sum(~isGood) > 5:
+                r_distance_predicted = np.hypot(x_predicted[-1] - x_predicted[0],
+                                                y_predicted[-1] - y_predicted[0])
+            else:
+                r_distance_predicted = 0
+
+            # Residuals from the polyfit are larger than 0.5 mm only in cases of
+            # catastrophic failure
+            if residuals_y[0] < residual_max and residuals_x[0] < residual_max\
+                    and ~((number_of_breaks == 1) and (r_distance_predicted > max_break_distance)):
```

Comment:
Put it all in parens, and you can get rid of the backslash. In general, I view backslashes as fragile, and do whatever I can to avoid them.